### PR TITLE
SI-7266 Stream leaks memory

### PR DESCRIPTION
**Pull request by Ichoran**
Changed tail-generation function to mutable and clear it after it's used to allow any captured memory to be freed once the tail has been generated.

(This is a case where a by-name parameter was used when a lazy val parameter was wanted instead.  If we ever get lazy val parameters, we should switch to that.)

### CODE REVIEW

File: src/library/scala/collection/immutable/Stream.scala [hunk 1]
```diff
@@ -1108,11 +1108,15 @@ object Stream extends SeqFactory[Stream] {
...
     @volatile private[this] var tlVal: Stream[A] = _
-    def tailDefined: Boolean = tlVal ne null
+    @volatile private[this] var tlGen = tl _
+    def tailDefined: Boolean = tlGen eq null
     override def tail: Stream[A] = {
       if (!tailDefined)
         synchronized {
-          if (!tailDefined) tlVal = tl
+          if (!tailDefined) {
+            tlVal = tlGen()
```
Comment (retronym) [inline src/library/scala/collection/immutable/Stream.scala:1117]: Couldn't another thread null out `tlGen` in the meantime? It's volatile, but we read it twice here.

Answering my own question: i suppose not, as we're in a `synchronized` section.

File: src/library/scala/collection/immutable/Stream.scala [hunk 2]
```diff
@@ -1108,11 +1108,15 @@ object Stream extends SeqFactory[Stream] {
     override def isEmpty = false
     override def head = hd
     @volatile private[this] var tlVal: Stream[A] = _
-    def tailDefined: Boolean = tlVal ne null
+    @volatile private[this] var tlGen = tl _
```
Comment (retronym) [inline src/library/scala/collection/immutable/Stream.scala:1111]: Confirmed that we don't retain `tl` as a field.

```
scala> :javap -private scala.collection.immutable.Stream$Cons
Compiled from "Stream.scala"
public final class scala.collection.immutable.Stream$Cons extends scala.collection.immutable.Stream{
    private final java.lang.Object hd;
    private volatile scala.collection.immutable.Stream tlVal;
    private volatile scala.Function0 tlGen;
    public boolean isEmpty();
    public java.lang.Object head();
    public boolean tailDefined();
    public scala.collection.immutable.Stream tail();
    public scala.collection.LinearSeqOptimized tail();
    public java.lang.Object tail();
    public scala.collection.immutable.Stream$Cons(java.lang.Object, scala.Function0);
}
```

File: src/library/scala/collection/immutable/Stream.scala (hunk 1 above)
Comment (Ichoran) [inline src/library/scala/collection/immutable/Stream.scala:1117]: That's the reason for the `synchronized` here.

File: src/library/scala/collection/immutable/Stream.scala (hunk 2 above)
Comment (Ichoran) [inline src/library/scala/collection/immutable/Stream.scala:1111]: Yeah, I checked the bytecode also.  One could also get rid of `tlGen` entirely by doing an `instanceof` to tell whether we were storing the generator of the value or the value itself, but this doesn't save any space at least on the Oracle JVM (object size increases 2, 4, ... Object fields.  1==0, 3==2, etc.

**Comment by retronym** [comment]
LGTM.

**Comment by gkossakowski** [comment]
LGTM as well.

**Pull request merged**